packages/contracts/contracts/test/plugin-mock/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol

```diff
@@ -0,0 +1,187 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.10;
+
+import {PermissionLib} from "../../../core/permission/PermissionLib.sol";
+import {PluginSetup} from "../../../plugin/PluginSetup.sol";
+import {IPluginSetup} from "../../../plugin/IPluginSetup.sol";
+import {PluginUUPSUpgradeableV1Mock, PluginUUPSUpgradeableV2Mock, PluginUUPSUpgradeableV3Mock} from "./PluginUUPSUpgradeableMock.sol";
+import {mockPermissions, mockHelpers, mockPluginProxy} from "../PluginMockData.sol";
+
+contract PluginUUPSUpgradeableSetupV1Mock is PluginSetup {
+    address internal pluginBase;
+
+    constructor() {
+        pluginBase = address(new PluginUUPSUpgradeableV1Mock());
+    }
+
+    /// @inheritdoc IPluginSetup
+    function prepareInstallationDataABI() external view virtual override returns (string memory) {
+        return "";
+    }
+
+    /// @inheritdoc IPluginSetup
+    function prepareInstallation(address _dao, bytes memory)
+        public
+        virtual
+        override
+        returns (
+            address plugin,
+            address[] memory helpers,
+            PermissionLib.ItemMultiTarget[] memory permissions
+        )
+    {
+        plugin = mockPluginProxy(pluginBase, _dao);
+        helpers = mockHelpers(1);
+        permissions = mockPermissions(0, 1, PermissionLib.Operation.Grant);
+    }
+
+    /// @inheritdoc IPluginSetup
+    function prepareUninstallationDataABI() external view virtual override returns (string memory) {
+        return "";
+    }
+
+    /// @inheritdoc IPluginSetup
+    function prepareUninstallation(
+        address _dao,
+        address _plugin,
+        address[] calldata _currentHelpers,
+        bytes calldata
+    ) external virtual override returns (PermissionLib.ItemMultiTarget[] memory permissions) {
+        (_dao, _plugin, _currentHelpers);
+        permissions = mockPermissions(0, 1, PermissionLib.Operation.Revoke);
+    }
+
+    /// @inheritdoc IPluginSetup
+    function getImplementationAddress() external view virtual override returns (address) {
+        return address(pluginBase);
+    }
+}
+
+contract PluginUUPSUpgradeableSetupV1MockBad is PluginUUPSUpgradeableSetupV1Mock {
+    function prepareInstallation(address _dao, bytes memory)
+        public
+        pure
+        override
+        returns (
+            address plugin,
+            address[] memory helpers,
+            PermissionLib.ItemMultiTarget[] memory permissions
+        )
+    {
+        (_dao);
+        plugin = address(0); // The bad behaviour is returning the same address over and over again
```

Comment:
IMO, getting the previously prepared plugin address from `_data` was more simulating the attach scenario.
Not sure about using address zero, but if previous test can pas then it is fine.

Reply:
Yes, the associated test was named `"reverts if plugin setup return the same address"` and is passing.